### PR TITLE
Add confirmations, blockheight, blockindex and blocktime to z_listreceivedbyaddress

### PR DESCRIPTION
**Pull request by oxarbitrage**
Fixes https://github.com/zcash/zcash/issues/3724

1- There was a PR to add confirmations to this call at https://github.com/zcash/zcash/pull/3836
I ported the commit from there and fixed test case by incrementing the confirmations as suggested at: https://github.com/zcash/zcash/pull/3836#issuecomment-499927807

2- Then added `blockheight`, `blockindex` and `blocktime`. To avoid some duplicated code (Sprout/Sapling) created a structure `trxblock`.

3- Original issue requests only time and blockindex however i think height is also important; if `blockindex` is the position of the transaction in the block then you are going to need also `height` to find it.

### CODE REVIEW

File: qa/rpc-tests/wallet_listreceived.py [hunk 1]
```diff
@@ -49,10 +49,22 @@ def run_test_release(self, release, height):
...
+        assert_equal(-1, r[0]['blockindex'])
+        assert_equal(0, r[0]['blockheight'])
 
         # Confirm transaction (1 ZEC from taddr to zaddr1)
         self.generate_and_sync(height+3)
 
+        # adjust previous result because now there is one more confirmation
+        r[0]['confirmations'] += 1
+
+        # adjust blockindex as now there are 2 transactions confirmed in the block
```
Comment (oxarbitrage) [inline qa/rpc-tests/wallet_listreceived.py]: This should actually be a question.

**Comment by holmesworcester** [comment]
FWIW, this would be helpful to us in our work on [Zbay](https://zbay.app).

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -3303,6 +3303,23 @@ CAmount getBalanceZaddr(std::string address, int minDepth = 1, bool ignoreUnspen
...
+    int height = 0;
+    int index = -1;
+    int64_t time = 0;
+
+    txblock(uint256 hash)
+    {
+        if (pwalletMain->mapWallet.count(hash)) {
+            const CWalletTx& wtx = pwalletMain->mapWallet[hash];
+            if (!wtx.hashBlock.IsNull())
+                height = mapBlockIndex[wtx.hashBlock]->nHeight;
```
Comment (str4d) [inline src/wallet/rpcwallet.cpp:3317]: These should not use the globals directly, instead having them passed in. But we can do that as part of a larger refactor.

**Comment by str4d** [comment]
@zkbot r+

File: qa/rpc-tests/wallet_nullifiers.py [hunk 3]
```diff
@@ -133,8 +133,10 @@ def run_test (self):
         assert_equal(myzaddr in self.nodes[3].z_listaddresses(), False)
         assert_equal(myzaddr in self.nodes[3].z_listaddresses(True), True)
 
-        # Node 3 should see the same received notes as node 2; however,
-        # some of the notes were change for node 2 but not for node 3.
+        # Node 3 should see the same received notes as node 2; however, there are 2 things:
+        # - Some of the notes were change for node 2 but not for node 3.
+        # - Each node wallet store transaction time as received. As
+        #   `wait_and_assert_operationid_status` is called node 2 and 3 are off by a few seconds.
         # Aside from that the recieved notes should be the same. So,
```
Comment (rex4539) [inline qa/rpc-tests/wallet_nullifiers.py:140]:
```suggestion
        # Aside from that the received notes should be the same. So,
```

Comment (rex4539) [reply qa/rpc-tests/wallet_nullifiers.py:140]: Combining the fix from https://github.com/zcash/zcash/pull/4456

**Comment by str4d** [comment]
@zkbot r+

**Pull request merged**